src/pyroboplan/planning/path_shortcutting.py

```diff
@@ -0,0 +1,149 @@
+
+
+def shortcut_path(
+    model, collision_model, q_path, max_iters=100, num_restarts=2, max_angle_step=0.05
+):
+    """
+    Performs path shortcutting by sampling two random points along the path and verifying
+    whether those two points can be connected directly without collisions.
+
+    This implementation is based on section 3.5.3 of:
+    https://motion.cs.illinois.edu/RoboticSystems/MotionPlanningHigherDimensions.html
+
+    Parameters
+    ----------
+        model : `pinocchio.Model`
+            The model of the robot.
+        collision_model : `pinocchio.Model`.
+            The model to use for collision checking.
+        q_path : list[array-like]
+            The list of joint configurations describing the path.
+        max_iters : int
+            The maximum iterations for randomly sampling shortcut points.
+        num_restarts : int
+            The number of restarts to try different random path shortcutting solutions.
+        max_angle_step : float
+            Maximum angle step, in radians, for collision checking along path segments.
+
+    Return
+    ------
+        list[array-like]
+            The shortest length path over all the shortcutting attempts.
+    """
+    # If the original path has 2 points or less, this is already a shortest path.
+    if len(q_path) < 3:
+        return q_path
+
+    best_path = q_path
+    best_path_length = get_path_length(best_path)
+
+    for _ in range(num_restarts + 1):
+        q_shortened = copy.deepcopy(q_path)
+        for _ in range(max_iters):
+            # If the path has been shortened to 2 points or less, this is already a shortest path.
+            if len(q_shortened) < 3:
+                break
+
+            # Sample two points along the path length
+            path_scalings = get_normalized_path_scaling(q_shortened)
+            while True:
+                low_point, high_point = sorted(np.random.random(2))
+                q_low, idx_low = get_configuration_from_normalized_path_scaling(
+                    q_shortened, path_scalings, low_point
+                )
+                q_high, idx_high = get_configuration_from_normalized_path_scaling(
+                    q_shortened, path_scalings, high_point
+                )
+                if idx_low < idx_high:
+                    break
+
+            # Check if the points are collision free. If they are, shortcut the path.
+            path_to_goal = discretize_joint_space_path([q_low, q_high], max_angle_step)
+            if not check_collisions_along_path(model, collision_model, path_to_goal):
+                q_shortened = (
+                    q_shortened[:idx_low] + [q_low, q_high] + q_shortened[idx_high:]
+                )
+
+        # Check if this is the best path so far
+        shortened_path_length = get_path_length(q_shortened)
+        if shortened_path_length < best_path_length:
+            best_path = copy.deepcopy(q_shortened)
```

Comment:
It doesn't _really_ matter but I don't think you need the deepcopy here - since you'll just reset `q_shortened` with a deep copy on the next iteration anyway.

Reply:
Seems like it wasn't needed in either plae!